### PR TITLE
Full IOMMU / VFIO support to ixy

### PR DESCRIPTION
**Pull request by huberste**
This pull request adds full IOMMU / VFIO support to ixy. It will compile without additional cmake flags and decide at runtime whether to use VFIO or not (by checking if the IOMMU is enabled and the VFIO driver is loaded for the device).
All initialized ixy devices (i.e. NICs) will be in the same VFIO container, thus be able to read from each others memory (needed for the forwarder).

### CODE REVIEW

File: src/pci.c [hunk 1]
```diff
@@ -52,10 +52,23 @@ uint8_t* pci_map_resource(const char* pci_addr) {
...
-	int fd = check_err(open(path, O_RDWR), "open pci resource");
+	int fd = check_err(open(path, flags), "open pci resource");
 	return fd;
 }
+
+/*
+ * Check if a file exist using stat() function
+ * return true if the file exist otherwise return false
+ */
+int fileexists(const char* filename){
```
Comment (emmericp) [inline src/pci.c]: used exactly once, better to inline for now

File: src/libixy-vfio.c [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+		error("Failed to set container. errno: %d", errno);
+		return -1;
+	}
+
+	if (firstsetup != 0) {
+		// Set vfio type (type1 is for IOMMU like VT-d or AMD-Vi) for the
+		// container.
+		// This can only be done after at least one group is in the container.
+		ret = ioctl(cfd, VFIO_SET_IOMMU, VFIO_TYPE1_IOMMU);
+		if(ret != 0){
```
Comment (emmericp) [inline src/libixy-vfio.c]: whitespace

File: src/libixy-vfio.c [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+		return ret;
+	}
+	if(!group_status.flags & VFIO_GROUP_FLAGS_VIABLE){
+		error("VFIO group is not viable - are all devices in the group bound to the VFIO driver?");
+		return -1;
+	}
+
+	// Add group to container
+	ret = ioctl(vfio_gfd, VFIO_GROUP_SET_CONTAINER, &cfd);
+	if(ret == -1){
```
Comment (emmericp) [inline src/libixy-vfio.c]: whitespace

File: src/libixy-vfio.c [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+		ret = ioctl(cfd, VFIO_SET_IOMMU, VFIO_TYPE1_IOMMU);
+		if(ret != 0){
+			error("Failed to set iommu type. errno: %d", -ret);
+			return -1;
+		}
+	}
+
+	// get device file descriptor
+	int vfio_fd = ioctl(vfio_gfd, VFIO_GROUP_GET_DEVICE_FD, pci_addr);
+	if(vfio_fd < 0){
```
Comment (emmericp) [inline src/libixy-vfio.c]: whitespace

File: src/libixy-vfio.c [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+	return vfio_fd;
+}
+
+/* returns a uint8_t pointer to the MMAPED region or MAP_FAILED if failed */
+uint8_t* vfio_map_region(int vfio_fd, int region_index) {
+	struct vfio_region_info region_info = { .argsz = sizeof(region_info) };
+	region_info.index = region_index;
+	int ret = ioctl(vfio_fd, VFIO_DEVICE_GET_REGION_INFO, &region_info);
+	if(ret == -1){
```
Comment (emmericp) [inline src/libixy-vfio.c]: whitespace

File: src/memory.c [hunk 6]
```diff
@@ -113,3 +141,13 @@ void pkt_buf_free(struct pkt_buf* buf) {
 	mempool->free_stack[mempool->free_stack_top++] = buf->mempool_idx;
 }
 
+// VFIO / IOMMU specific stuff
+/* reads the global VFIO container */
```
Comment (emmericp) [inline src/memory.c]: prefer C++-style comments

File: src/memory.c [hunk 7]
```diff
@@ -113,3 +141,13 @@ void pkt_buf_free(struct pkt_buf* buf) {
...
+int get_vfio_container() {
+	return VFIO_CONTAINER_FILE_DESCRIPTOR;
+}
+
+/* globally sets the VFIO container and returns the set value */
```
Comment (emmericp) [inline src/memory.c]: C++-style comments, please

File: src/memory.h [hunk 8]
```diff
@@ -49,4 +51,11 @@ uint32_t pkt_buf_alloc_batch(struct mempool* mempool, struct pkt_buf* bufs[], ui
 struct pkt_buf* pkt_buf_alloc(struct mempool* mempool);
 void pkt_buf_free(struct pkt_buf* buf);
 
+// VFIO / IOMMU specific stuff
+/* reads the global VFIO container */
```
Comment (emmericp) [inline src/memory.h]: C++-style comments

File: src/memory.h [hunk 9]
```diff
@@ -49,4 +51,11 @@ uint32_t pkt_buf_alloc_batch(struct mempool* mempool, struct pkt_buf* bufs[], ui
...
+int get_vfio_container();
+
+/* globally sets the VFIO container */
```
Comment (emmericp) [inline src/memory.h]: C++-style comments

File: src/libixy-vfio.c [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+	struct vfio_region_info conf_reg = { .argsz = sizeof(conf_reg) };
+	conf_reg.index = VFIO_PCI_CONFIG_REGION_INDEX;
+	check_err(ioctl(device_fd, VFIO_DEVICE_GET_REGION_INFO, &conf_reg), "get vfio config region info");
+	uint16_t dma = 0;
+	assert(pread(device_fd, &dma, 2, conf_reg.offset + command_register_offset) == 2);
+	dma |= 1 << bus_master_enable_bit;
+	assert(pwrite(device_fd, &dma, 2, conf_reg.offset + command_register_offset) == 2);
+}
+
+/* returns the devices file descriptor or -1 on error */
```
Comment (emmericp) [inline src/libixy-vfio.c]: C++-style comments

File: src/libixy-vfio.c [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+	region_info.index = region_index;
+	int ret = ioctl(vfio_fd, VFIO_DEVICE_GET_REGION_INFO, &region_info);
+	if(ret == -1){
+		// Failed to set iommu type
+		return MAP_FAILED; // MAP_FAILED == ((void *) -1)
+	}
+	return (uint8_t*) check_err(mmap(NULL, region_info.size, PROT_READ | PROT_WRITE, MAP_SHARED, vfio_fd, region_info.offset), "mmap vfio bar0 resource");
+}
+
+/* returns iova (physical address of the DMA memory from device view) on success */
```
Comment (emmericp) [inline src/libixy-vfio.c]: C++-style comments

File: src/libixy-vfio.c [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+		.iova = iova,
+		.size = size < MIN_DMA_MEMORY ? MIN_DMA_MEMORY : size,
+		.argsz = sizeof(dma_map),
+		.flags = VFIO_DMA_MAP_FLAG_READ | VFIO_DMA_MAP_FLAG_WRITE};
+	int cfd = get_vfio_container();
+	check_err(ioctl(cfd, VFIO_IOMMU_MAP_DMA, &dma_map), "IOMMU Map DMA Memory");
+	return iova;
+}
+
+/* unmaps previously mapped DMA region. returns 0 on success */
```
Comment (emmericp) [inline src/libixy-vfio.c]: C++-style comments

File: src/libixy-vfio.h [hunk 13]
```diff
@@ -0,0 +1,26 @@
+#ifndef LIBIXY_VFIO_H
+#define LIBIXY_VFIO_H
+
+#include <stdint.h>
+
```
Comment (emmericp) [inline src/libixy-vfio.h:5]: C++-style comments in the whole file unless multiple lines are needed for a function description

**Review by emmericp — CHANGES_REQUESTED** [review]

File: src/driver/ixgbe.c [hunk 14]
```diff
@@ -282,8 +292,24 @@ struct ixy_device* ixgbe_init(const char* pci_addr, uint16_t rx_queues, uint16_t
...
+	// Allocate memory for the ixgbe device that will be returned
 	struct ixgbe_device* dev = (struct ixgbe_device*) malloc(sizeof(struct ixgbe_device));
 	dev->ixy.pci_addr = strdup(pci_addr);
+
+	// Check if we want the VFIO stuff
+	// This is done by checking if the device is in an IOMMU group.
+	char path[PATH_MAX];
+	snprintf(path, PATH_MAX, "/sys/bus/pci/devices/%s/iommu_group", pci_addr);
+	struct stat buffer;
+	dev->ixy.vfio = stat(path,&buffer) == 0;
```
Comment (emmericp) [inline src/driver/ixgbe.c]: space missing

File: src/libixy-vfio.c [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+	// bit 2 is "bus master enable", see PCIe 3.0 specification section 7.5.1.1
+	int bus_master_enable_bit = 2;
+	// Get region info for config region
+	struct vfio_region_info conf_reg = {.argsz = sizeof(conf_reg)};
+	conf_reg.index = VFIO_PCI_CONFIG_REGION_INDEX;
+	check_err(ioctl(device_fd, VFIO_DEVICE_GET_REGION_INFO, &conf_reg),
+			  "get vfio config region info");
+	uint16_t dma = 0;
+	assert(pread(device_fd, &dma, 2,
+				 conf_reg.offset + command_register_offset) == 2);
```
Comment (emmericp) [inline src/libixy-vfio.c]: merge into a single line, please

File: src/libixy-vfio.c [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+	dma |= 1 << bus_master_enable_bit;
+	assert(pwrite(device_fd, &dma, 2,
+				  conf_reg.offset + command_register_offset) == 2);
```
Comment (emmericp) [inline src/libixy-vfio.c]: merge into a single line, please

File: src/libixy-vfio.c [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+		error("Failed to convert group id '%s' to int", group_name);
+		return -1;
+	}
+
+	int firstsetup = 0; // Need to set up the container exactly once
+	int cfd = get_vfio_container();
+	if (cfd == -1) {
+		firstsetup = 1;
+		// open vfio file to create new cfio conainer
+		cfd = open("/dev/vfio/vfio", O_RDWR);
```
Comment (emmericp) [inline src/libixy-vfio.c]: is this world-writable by default or do we need an additional chown in the setup instructions?

File: src/libixy-vfio.c [hunk 18]
```diff
@@ -0,0 +1,189 @@
...
+	region_info.index = region_index;
+	int ret = ioctl(vfio_fd, VFIO_DEVICE_GET_REGION_INFO, &region_info);
+	if (ret == -1) {
+		// Failed to set iommu type
+		return MAP_FAILED; // MAP_FAILED == ((void *) -1)
+	}
+	return (uint8_t*) check_err(mmap(NULL, region_info.size,
+									 PROT_READ | PROT_WRITE, MAP_SHARED,
+									 vfio_fd, region_info.offset),
+								"mmap vfio bar0 resource");
```
Comment (emmericp) [inline src/libixy-vfio.c]: this formatting looks wrong with both tab size 4 and 8; I'd probably just merge it into one line?

File: src/libixy-vfio.c [hunk 19]
```diff
@@ -0,0 +1,189 @@
...
+}
+
+// returns iova (physical address of the DMA memory from device view) on success
+uint64_t vfio_map_dma(void* vaddr, uint32_t size) {
+	uint64_t iova =
+		(uint64_t) vaddr; // Identity mapping makes sense here because why not
```
Comment (emmericp) [inline src/libixy-vfio.c]: single line

Comment (emmericp) [reply src/libixy-vfio.c]: should this be called identity mapping? it maps the iova to the normal virtual address? wouldn't identity mapping be iova == physical address?

maybe change comment to "map iova to our virtual address"?

File: src/libixy-vfio.c [hunk 20]
```diff
@@ -0,0 +1,189 @@
...
+	return iova;
+}
+
+// unmaps previously mapped DMA region. returns 0 on success
+uint64_t vfio_unmap_dma(int fd, uint64_t iova, uint32_t size) {
+	struct vfio_iommu_type1_dma_unmap dma_unmap = {
+		.argsz = sizeof(dma_unmap),
+		.flags = VFIO_DMA_MAP_FLAG_READ | VFIO_DMA_MAP_FLAG_WRITE,
+		.iova = iova,
+		.size = size};
```
Comment (emmericp) [inline src/libixy-vfio.c]: closing brace on new line

File: src/memory.c [hunk 21]
```diff
@@ -31,33 +42,46 @@ static uint32_t huge_pg_id;
...
+	if (VFIO_CONTAINER_FILE_DESCRIPTOR != -1) {
+		// VFIO == -1 means that there is no VFIO container set, i.e. VFIO / IOMMU is not activated
+		debug("allocating dma memory via VFIO");
+		void* virt_addr = (void*) check_err(mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_SHARED | MAP_ANONYMOUS | MAP_HUGETLB | MAP_HUGE_2MB, -1, 0), "mmap hugepage");
+		// create IOMMU mapping
+		uint64_t iova = (uint64_t) vfio_map_dma(virt_addr, size);
+		return (struct dma_memory){
+			// for VFIO, this needs to point to the device view memory = IOVA!
+			.virt = virt_addr,
+			.phy = iova};
```
Comment (emmericp) [inline src/memory.c]: closing brace to next line

File: src/memory.c [hunk 22]
```diff
@@ -68,21 +92,27 @@ struct mempool* memory_allocate_mempool(uint32_t num_entries, uint32_t entry_siz
...
 	// require entries that neatly fit into the page size, this makes the memory pool much easier
 	// otherwise our base_addr + index * size formula would be wrong because we can't cross a page-boundary
-	if (HUGE_PAGE_SIZE % entry_size) {
+	if ((VFIO_CONTAINER_FILE_DESCRIPTOR == -1) && HUGE_PAGE_SIZE % entry_size) {
 		error("entry size must be a divisor of the huge page size (%d)", HUGE_PAGE_SIZE);
 	}
 	struct mempool* mempool = (struct mempool*) malloc(sizeof(struct mempool) + num_entries * sizeof(uint32_t));
-	struct dma_memory mem = memory_allocate_dma(num_entries * entry_size, false);
+	struct dma_memory mem;
+	mem = memory_allocate_dma(num_entries * entry_size, false);
```
Comment (emmericp) [inline src/memory.c]: undo this change please

File: src/memory.c [hunk 23]
```diff
@@ -113,3 +143,8 @@ void pkt_buf_free(struct pkt_buf* buf) {
 	mempool->free_stack[mempool->free_stack_top++] = buf->mempool_idx;
 }
 
+// reads the global VFIO container
+int get_vfio_container() { return VFIO_CONTAINER_FILE_DESCRIPTOR; }
```
Comment (emmericp) [inline src/memory.c]: newlines, please

File: src/memory.c [hunk 24]
```diff
@@ -113,3 +143,8 @@ void pkt_buf_free(struct pkt_buf* buf) {
...
+
+// globally sets the VFIO container and returns the set value
+void set_vfio_container(int fd) { VFIO_CONTAINER_FILE_DESCRIPTOR = fd; }
```
Comment (emmericp) [inline src/memory.c]: newlines, please

File: src/memory.h [hunk 25]
```diff
@@ -1,14 +1,16 @@
...
 #include <stddef.h>
-#include <assert.h>
+#include <stdint.h>
+#include <unistd.h>
+
+struct ixy_device;
 
 #define HUGE_PAGE_BITS 21
-#define HUGE_PAGE_SIZE (1 << HUGE_PAGE_BITS)
+#define HUGE_PAGE_SIZE (1 << HUGE_PAGE_BITS) // 2.097.152 = 2M
```
Comment (emmericp) [inline src/memory.h]: use _ or , as separator, please

File: README.md [hunk 26]
```diff
@@ -87,6 +87,35 @@ We currently have a simple check if the device is actually a NIC, but trying to
 	
 	which means that I have to pass `0000:03:00.0` as parameter to use it.
 
+## Using the IOMMU / VFIO
+The usage of the IOMMU via the `vfio-pci` driver is implemented for ixge devices (Intel X520, X540 and X550).
```
Comment (emmericp) [inline README.md]: *ixgbe

oxford comma, please

File: README.md [hunk 27]
```diff
@@ -87,6 +87,35 @@ We currently have a simple check if the device is actually a NIC, but trying to
...
+To use it, you have to:
+
+0. Enable the IOMMU in the BIOS.
+	On most Intel machines, the BIOS entry is called `VT-d` and has to be enabled additionally to any other virtualization techniques.
```
Comment (emmericp) [inline README.md]: *in addition to (...) technique.

File: README.md [hunk 28]
```diff
@@ -87,6 +87,35 @@ We currently have a simple check if the device is actually a NIC, but trying to
...
+
+1. Enable the IOMMU in the linux kernel.
+	Add `intel_iommu=on` to your cmdline (if you are running a grub, the file /etc/default/grub.cfg contains a `GRUB_CMDLINE_LINUX` where you can add it).
```
Comment (emmericp) [inline README.md]: path in `backticks`

File: src/libixy-vfio.c [hunk 29]
```diff
@@ -0,0 +1,189 @@
...
+	if (vfio_gfd < 0) {
+		error("Failed to open vfio group");
+		return -1;
+	}
+
+	// check if group is viable
+	struct vfio_group_status group_status = {.argsz = sizeof(group_status)};
+	ret = ioctl(vfio_gfd, VFIO_GROUP_GET_STATUS, &group_status);
+	if (ret == -1) {
+		error("failed to get VFIO group status. errno: %d", errno);
```
Comment (emmericp) [inline src/libixy-vfio.c]: whole file: printing error numbers without `strerror` is horrible user experience. can this use `check_err` instead?

**Review by emmericp — CHANGES_REQUESTED** [review]

File: src/libixy-vfio.c (hunk 17 above)
Comment (huberste) [reply src/libixy-vfio.c]: a) it's world-writeable
b) it's already in the README

File: src/libixy-vfio.c [hunk 30]
```diff
@@ -0,0 +1,141 @@
...
+		firstsetup = 1;
+		// open vfio file to create new vfio container
+		cfd = check_err(open("/dev/vfio/vfio", O_RDWR), "open /dev/vfio/vfio");
+		set_vfio_container(cfd);
+
+		// check if the container's API version is the same as the VFIO API's
+		check_err((ioctl(cfd, VFIO_GET_API_VERSION) == VFIO_API_VERSION) - 1, "get a valid API version from the container");
+
+		// check if type1 is supported
+		check_err((ioctl(cfd, VFIO_CHECK_EXTENSION, VFIO_TYPE1_IOMMU) == 1) - 1, "get Type1 IOMMU support from the IOMMU container");
```
Comment (emmericp) [inline src/libixy-vfio.c:69]: please don't do that magic -1, that's not very readable

File: src/libixy-vfio.c [hunk 31]
```diff
@@ -0,0 +1,141 @@
...
+	}
+
+	// open VFIO group containing the device
+	snprintf(path, sizeof(path), "/dev/vfio/%d", groupid);
+	int vfio_gfd = check_err(open(path, O_RDWR), "open vfio group");
+
+	// check if group is viable
+	struct vfio_group_status group_status = {.argsz = sizeof(group_status)};
+	check_err(ioctl(vfio_gfd, VFIO_GROUP_GET_STATUS, &group_status), "get VFIO group status");
+	check_err(((group_status.flags & VFIO_GROUP_FLAGS_VIABLE) > 0) - 1, "get viable VFIO group - are all devices in the group bound to the VFIO driver?");
```
Comment (emmericp) [inline src/libixy-vfio.c:79]: same

**Pull request merged**